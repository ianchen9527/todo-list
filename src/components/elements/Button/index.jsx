import React from "react";
import styles from "./index.module.scss";
import PropTypes from "prop-types";
import cx from "classnames";

const getWrapperClassName = (disabled) =>
  cx({
    [styles["wrapper"]]: true,
    [styles["disabled"]]: disabled,
  });

const Button = ({ icon, loading, onClick, text, disabled }) => {
  return (
    <div className={getWrapperClassName(disabled)} onClick={onClick}>
      {icon && <div className={styles[`icon-${icon}`]} />}
      {text}
      {loading && <div className={styles["loading"]} />}
    </div>
  );
};

Button.propTypes = {
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  icon: PropTypes.string,
  text: PropTypes.string,
};

Button.defaultProps = {
  onClick: null,
  diabled: false,
  loading: false,
  icon: "add",
  text: "",
};

export default Button;
