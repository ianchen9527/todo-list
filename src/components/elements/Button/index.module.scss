@import "../../../stylesheets/constants/colors.scss";

.wrapper {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-content: center;
  height: 40px;
  background-color: $button;
  border: 1px solid $black;
  border-radius: 5px;
  cursor: pointer;
  color: $black;
  font-family: "Roboto";
  font-size: 14px;
  line-height: 21px;

  &:hover {
    background-color: $button-hover;
  }

  &.disabled {
    opacity: 0.5;
    cursor: default;

    &:hover {
      background-color: $button;
    }
  }

  @mixin icon {
    width: 20px;
    height: 20px;
    display: block;
    background-size: 20px 20px;
    margin-right: 15px;
  }

  .icon-add {
    @include icon();
    
    background-image: url("../../../assets/images/add.svg");
  }

  .icon-edit {
    @include icon();
    
    background-image: url("../../../assets/images/edit.svg");
  }

  .loading {
    width: 20px;
    height: 20px;
    display: block;
    background-size: 20px 20px;
    background-image: url("../../../assets/images/loading.gif");
    margin-left: 15px;
  }
}
