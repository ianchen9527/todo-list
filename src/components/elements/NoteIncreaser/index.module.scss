@import "../../../stylesheets/constants/colors.scss";

.wrapper {
  width: 200px;
  height: 40px;
}
