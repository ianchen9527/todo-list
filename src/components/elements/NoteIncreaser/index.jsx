import React from "react";
import PropTypes from "prop-types";
import styles from "./index.module.scss";
import Button from "../Button";

const NoteIncreaser = (props) => {
  return <div className={styles['wrapper']}>
    <Button {...props} text="New Note" icon="add" />
    </div>;
};

NoteIncreaser.propTypes = {
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
};

NoteIncreaser.defaultProps = {
  onClick: null,
  diabled: false,
  loading: false,
};

export default NoteIncreaser;
