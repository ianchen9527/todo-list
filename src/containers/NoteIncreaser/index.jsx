import { compose } from "redux";
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";
import {
  getDisabled,
  getLoading,
  getOnClick,
} from "../../selectors/containers/noteIncreaser";
import NoteIncreaser from "../../components/elements/NoteIncreaser";

const mapStateToProps = (state, { match }) => ({
  disabled: getDisabled(match),
  loading: getLoading(state),
});

const mapDispatchToProps = (dispatch, { history, match }) => ({
  onClick: getOnClick(dispatch, history, match),
});

export default compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps)
)(NoteIncreaser);
