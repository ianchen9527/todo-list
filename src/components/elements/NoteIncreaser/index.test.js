import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import NoteIncreaser from "./index";
import Button from "../Button";

describe("NoteIncreaser", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<NoteIncreaser />);
  });

  describe("default props", () => {
    it("should render without crashing", () => {
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe("props.loading", () => {
    let loading;
    describe("is true", () => {
      beforeEach(() => {
        loading = true;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });

    describe("is false", () => {
      beforeEach(() => {
        loading = false;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });
  });

  describe("props.disabled", () => {
    let disabled;
    describe("is true", () => {
      beforeEach(() => {
        disabled = true;
        wrapper.setProps({ disabled });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });

    describe("is false", () => {
      beforeEach(() => {
        disabled = false;
        wrapper.setProps({ disabled });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });
  });

  describe("props.onClick", () => {
    let onClick;
    describe("is set", () => {
      beforeEach(() => {
        onClick = jest.fn();
        wrapper.setProps({ onClick });
      });

      it("should be passed to the Button", () => {
        wrapper.find(Button).props().onClick();
        expect(onClick).toHaveBeenCalled();
      });
    });
  });
});
