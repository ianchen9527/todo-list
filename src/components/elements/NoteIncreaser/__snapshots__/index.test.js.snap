// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteIncreaser default props should render without crashing 1`] = `
<div
  className="wrapper"
>
  <Button
    disabled={false}
    icon="add"
    loading={false}
    onClick={null}
    text="New Note"
  />
</div>
`;

exports[`NoteIncreaser props.disabled is false should render without crashing 1`] = `
<div
  className="wrapper"
>
  <Button
    disabled={false}
    icon="add"
    loading={false}
    onClick={null}
    text="New Note"
  />
</div>
`;

exports[`NoteIncreaser props.disabled is true should render without crashing 1`] = `
<div
  className="wrapper"
>
  <Button
    disabled={true}
    icon="add"
    loading={false}
    onClick={null}
    text="New Note"
  />
</div>
`;

exports[`NoteIncreaser props.loading is false should render without crashing 1`] = `
<div
  className="wrapper"
>
  <Button
    disabled={false}
    icon="add"
    loading={false}
    onClick={null}
    text="New Note"
  />
</div>
`;

exports[`NoteIncreaser props.loading is true should render without crashing 1`] = `
<div
  className="wrapper"
>
  <Button
    disabled={false}
    icon="add"
    loading={true}
    onClick={null}
    text="New Note"
  />
</div>
`;
